src/backend/distributed/executor/multi_utility.c

```diff
@@ -883,6 +897,223 @@ ProcessAlterObjectSchemaStmt(AlterObjectSchemaStmt *alterObjectSchemaStmt,
 
 
 /*
+ * ProcessVacuumStmt processes vacuum statements that may need propagation to
+ * distributed tables. If a VACUUM or ANALYZE command references a distributed
+ * table, it is propagated to all involved nodes; otherwise, this function will
+ * immediately exit after some error checking.
+ *
+ * Unlike other Process functions within this file, this function does not
+ * return a modified parse node, as it is expected that the local VACUUM or
+ * ANALYZE has already been processed.
+ */
+static void
+ProcessVacuumStmt(VacuumStmt *vacuumStmt, const char *vacuumCommand)
+{
+	Oid relationId = InvalidOid;
+	List *taskList = NIL;
+	bool supportedVacuumStmt = false;
+
+	if (vacuumStmt->relation != NULL)
+	{
+		relationId = RangeVarGetRelid(vacuumStmt->relation, NoLock, false);
+	}
+
+	supportedVacuumStmt = IsSupportedDistributedVacuumStmt(relationId, vacuumStmt);
+	if (!supportedVacuumStmt)
+	{
+		return;
+	}
+
+	taskList = VacuumTaskList(relationId, vacuumStmt);
+
+	SavedMultiShardCommitProtocol = MultiShardCommitProtocol;
+	MultiShardCommitProtocol = COMMIT_PROTOCOL_BARE;
+	ExecuteModifyTasksWithoutResults(taskList);
+}
+
+
+/*
+ * IsSupportedDistributedVacuumStmt returns whether distributed execution of a
+ * given VacuumStmt is supported. The provided relationId (if valid) represents
+ * the table targeted by the provided statement.
+ *
+ * Returns true if the statement requires distributed execution and returns
+ * false otherwise; however, this function will raise errors if the provided
+ * statement needs distributed execution but contains unsupported options.
+ */
+static bool
+IsSupportedDistributedVacuumStmt(Oid relationId, VacuumStmt *vacuumStmt)
+{
+	const char *stmtName = (vacuumStmt->options & VACOPT_VACUUM) ? "VACUUM" : "ANALYZE";
+
+	if (vacuumStmt->relation == NULL && EnableDDLPropagation)
+	{
+		/* WARN and exit early for local unqualified VACUUM commands */
+		ereport(WARNING, (errmsg("not propagating %s command to worker nodes", stmtName),
+						  errhint("Provide a specific table in order to %s "
+								  "distributed tables.", stmtName)));
+
+		return false;
+	}
+
+	if (!OidIsValid(relationId) || !IsDistributedTable(relationId))
+	{
+		return false;
+	}
+
+	if (!EnableDDLPropagation)
+	{
+		/* WARN and exit early if DDL propagation is not enabled */
+		ereport(WARNING, (errmsg("not propagating %s command to worker nodes", stmtName),
+						  errhint("Set citus.enable_ddl_propagation to true in order to "
+								  "send targeted %s commands to worker nodes.",
+								  stmtName)));
+	}
```

Comment:
The logic here is ugly, I know, but basically:

  * If a user runs `VACUUM` (i.e. against all tables) and has `EnableDDLPropagation` set to _on_, the local `VACUUM` runs, but we'll warn that it's not being propagated anywhere since that's unsupported
  * If a user runs `VACUUM distTable` and has `EnableDDLPropagation` set to _off_, it seemed worthwhile to warn the user in case they erroneously had the GUC on the wrong setting and honestly thought they were going to get a distributed `VACUUM`